draft: Use a SymPy Array not a Matrix for non-Expr

cbm755 opened the pull request:
It has mostly the same semantics as Matrix but can contain more general
things.  There might be some issues about different shapes of empties
though...  WIP on a fix for #1052 and #1055.

cbm755 committed:
Use a SymPy Array not a Matrix for non-Expr

It has mostly the same semantics as Matrix but can contain more general
things.  There might be some issues about different shapes of empties
though...  WIP on a fix for #1052.

cbm755 commented:
@alexvong1995 here's a half-baked idea for Issue #1124: roughly when we have non-`Expr` to be put in a `Matrix`, we instead make an `Array`.  This will be important for future SymPy as non-Expr in Matrix is deprecated.

cbm755 commented:
@alexvong1995 What do you think of this?  Basically in a few places (probably not that many) we return Array instead of Matrix.  Specifically we return Matrix greedily whenever everything is a Expr, else Array.

(It will need fixes upstream but I'm less worried about that.)

Maybe you can try doing this in the `mat_rclist_` and other helper functions, just to get a feel for the idea.

alexvong1995 commented:
I think the the test failure is partly due to `dbout` being define in `inst/private/python_header.py` (for non-Pythonic IPC) but not in `inst/private/python_ipc_native.m` (for Pythonic IPC).

I'm not sure what to do about it. It seems there're more code in `inst/private/python_header.py` than in `inst/private/python_ipc_native.m`.

cbm755 commented:
that could be true: I haven't looked at pythonic much recently.  You can always comment out "dbout()" (it is just debugging).  With Pythonic you can probably just `print()` instead.

Anyway, I think even concentrating on `system`-based, this idea is not complete.

Commit:
@sym/private/mat_rclist_access.m: Test Array-compatible code.

See #1194 for more information.

* inst/@sym/private/mat_rclist_access.m: Test it.

Commit:
@sym/private/mat_rclist_asgn.m: Test Array-compatible code.

See #1194 for more information.

* inst/@sym/private/mat_rclist_asgn.m: Test it.

Commit:
Python header: Add new function 'make_matrix_or_array'.

Fixes #1211.

* inst/private/python_header.py: Add it.
* inst/private/python_ipc_native.m: Add it.

cbm755 commented:
@alexvong1995 you can rebase and force push this on top of main (if you like).

alexvong1995 commented:
I think what we should do next is to make `@sym/horzcat` and `@sym/vertcat` Array-compatible. I haven't done any tests yet but I suspect all those `[t t; f f]` where `t` is `sym(true)` and `f` is `sym(false)` are causing many test failures.

Commit:
@sym/vertcat: Make function Array-compatible.

* inst/@sym/vertcat.m: Implement it.

alexvong1995 commented:
------- Original Message -------
On Monday, August 29th, 2022 at 12:04 AM, Colin B. Macdonald ***@***.***> wrote:


> @alexvong1995 you can rebase and force push this on top of main (if you like).
> 

rebased onto main and force pushed as a7c44c0d78bd6be66b6795ed339751657b1c3a3d

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

Commit:
@sym/{horzcat,vertcat}: Fix typo.

* inst/@sym/{horzcat,vertcat}.m: Fix them.

Commit:
@sym/horzcat: Make function Array-compatible by simplifying it.

* inst/@sym/horzcat.m: Simplify it.

Commit:
@sym/transpose: Make function Array-compatible.

This is a prerequisite for making @sym/horzcat Array-compatible.

Partially fixes <https://github.com/cbm755/octsympy/issues/1215>.

* inst/@sym/transpose.m: Implement it.

Commit:
Python header: Rename 'make_matrix_or_array' -> 'make_2d_sym'.

Generalises 'make_matrix_or_array' so that we can choose to represent
non-Matrix 2D sym to be something other than an Array in the
future. For example, we could use TableForm.

* inst/private/{python_header.py,python_ipc_native.m}: Rename function
'make_matrix_or_array' -> 'make_2d_sym', variable 'dbg_no_array' ->
'dbg_matrix_only' and adjust accordingly.
* inst/@sym/private/mat_rclist_{access,asgn}.m: Adjust accordingly.
* inst/@sym/vertcat.m: Adjust accordingly.

Commit:
@sym/private/mat_rclist_{access,asgn}: Remove unused variables.

* inst/@sym/private/mat_rclist_{access,asgn}.m: Remove them.

Commit:
@sym/repmat: Re-implement function without using 'pycall_sympy__'.

The function is now compatible with non-Matrix 2D sym (e.g. Array,
TableForm). Also, it now outputs empty matrices in the same way as
upstream Octave (fixes #1218).

* inst/@sym/repmat.m: Re-implement it and add tests accordingly.

cbm755 commented on inst/@sym/horzcat.m:
```diff
@@ -45,25 +46,8 @@
-  end
-  h = pycall_sympy__ (cmd, varargin{:});
+  args = cellfun (@transpose, varargin, 'UniformOutput', false);
+  h = vertcat (args{:}).';
```
while I think of it: I approve of this kind of "surface decrease" (having fewer functions that call `pycall_sympy__`) BUT it is important to be aware that there is a downside: namely more "roundtrips" between Python and Octave.  This is most noticeable with `ipc-system` but even with popen2, there are additional copies being made as everything is serialized back-and-forth over the pipe.

Its analogous to HTTP where all objects keep getting converted back-and-forth to JSON.

No action required but its something to keep in mind: perfect software design here might blow-up the latency and IO costs.

alexvong1995 replied:
Interesting thoughts!

Normally when working with a single language, the classic solution is to use an (optimizing) compiler to optimize away all the redundant conversions. Unfortunately, this doesn't work for us because of language barrier: a compiler can't optimize across language boundary :(

---

Some of my thoughts:

First, pythonic may have already solved our problem! While I am not familiar with the internals of pythonic, perhaps using pythonic's `py.`... instead of `pycall_sympy__` can avoid the needless conversions when going through a pipe. Of course, we should benchmark to see if it's indeed the case.

I believe this was the original *future* subproject but we changed it to the current (more urgent) one.

---

Another way is to create a octsmypy python library which contains one python function for each m-file function.

For example, say we want to implement m-file function `@sym/transpose`, `@sym/vertcat` and `@sym/horzcat`. Then our octsmypy python library should contain them as well. More precisely, we can do:

```
from octsympy.sym import transpose, vertcat, horzcat
```

To implement horzcat in terms of vertcat and transpose, we do it inside the octsmypy python library:

```
def horzcat(*XX):
    return transpose(vertcat(*[transpose(X) for X in XX]))
```

without the redundant conversions.
    
Finally, the m-file functions `@sym/transpose`, `@sym/vertcat` and `@sym/horzcat` are now thin wrappers over the corresponding octsympy python library functions, such as:

```
function h = vertcat (varargin)
    args = ...
    h = pycall_sympy__ ('from return octsympy.sym.vertcat(*_ins)', args{:});
end
```

and so on...

Perhaps this can be mentioned in the final report as possible future projects (?)

cbm755 replied:
I agree Pythonic offers possibilities.  I think it currently still does conversion: a 'sym' is really just a struct of the 'srepr' (string representation) from SymPy, indep of IPC mechanism.  But your 'xsym' idea would change that.

Agree about mentioing these things ad future projects.

Another thing to mention: CI pythonic runs are twice as slow as Popen2.  I susect there migit be some 'low hanging fruit' in the Pythonic project.

cbm755 replied:
And just for the record: another alternative that requires very little effort is (a) being mindful of this issue and aiming for 1 (or a small number) of language crosses per function (b) balancing that against technical debt of code duplication etc.  This is the approach we've been using so far.

cbm755 commented:
I think there was a way to turn an warning into an exception in Python.  That might be useful when debugging here...  like `SymPyDeprecationWarning` becomes an exception.

cbm755 committed:
elementwise_op: use 2d loop for Array

Array linear indexing is not the same as Matrix linear indexing.
Fixes #1222.

cbm755 committed:
uniop_bool_helper: Array support in "bool" case

This routine is supposed to return a logical Octave array.  Do linear
indexing on the flattened tolist output which will work on both Array
and Matrix.  Fixes #1221.

cbm755 committed:
subs: use make_matrix_or_array and use 2-d indexing

Fixes Issue #1226.  Related to Issue #1222.

cbm755 committed:
logical: remove 1-d indexing for Array support

cbm755 committed:
Minor fix to test: ctranspose of logical is tested elsewhere

In fact it is failing elsewhere but the point of *this* test is to
change the orientation of a vector.

cbm755 commented on inst/@sym/private/elementwise_op.m:
```diff
@@ -97,7 +97,10 @@
           'elements = []'
           'for i in range(0, len(q)):'
           '    elements.append(_op(*[k[i] if isinstance(k, MatrixBase) else k for k in _ins]))'
```
I'm not sure what test hits this code on line 99 yet, but I think from doing #1226 that this will need to be replaced with either 2-d indexing or some kind of `flatten()`

cbm755 replied:
wait, I think I already did this in #1223...  :-X

cbm755 committed:
Oops MatrixBase

cbm755 committed:
isAlways: avoid 1-d matrix/Array indexing

alexvong1995 commented:
------- Original Message -------
On Friday, September 2nd, 2022 at 4:57 PM, Colin B. Macdonald ***@***.***> wrote:


> I think there was a way to turn an warning into an exception in Python. That might be useful when debugging here... like `SymPyDeprecationWarning` becomes an exception.
>
The following example works for me:


--8<---------------cut here---------------start------------->8---
from sympy import Matrix
from sympy.utilities.exceptions import SymPyDeprecationWarning
import warnings


warnings.filterwarnings("error", category=SymPyDeprecationWarning)
Matrix([[True]])
--8<---------------cut here---------------end--------------->8---


I figure this out after reading multiple issues https://github.com/sympy/sympy/issues?q=is%3Aissue+is%3Aopen+SymPyDeprecationWarning of sympy. It would be good to know if this is documented somewhere or not.


> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

alexvong1995 commented:
------- Original Message -------
On Friday, September 2nd, 2022 at 4:57 PM, Colin B. Macdonald ***@***.***> wrote:


> I think there was a way to turn an warning into an exception in Python. That might be useful when debugging here... like `SymPyDeprecationWarning` becomes an exception.
> 

The following example works for me:

--8<---------------cut here---------------start------------->8---
from sympy import Matrix
from sympy.utilities.exceptions import SymPyDeprecationWarning
import warnings

warnings.filterwarnings("error", category=SymPyDeprecationWarning)
Matrix([[True]])
--8<---------------cut here---------------end--------------->8---

I figure this out after reading multiple issues https://github.com/sympy/sympy/issues?q=is%3Aissue+is%3Aopen+SymPyDeprecationWarning of sympy. It would be good to know if this is documented somewhere or not.

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

alexvong1995 commented:
------- Original Message -------
On Friday, September 2nd, 2022 at 4:57 PM, Colin B. Macdonald ***@***.***> wrote:


> I think there was a way to turn an warning into an exception in Python. That might be useful when debugging here... like `SymPyDeprecationWarning` becomes an exception.
> 

# The following example works for me:


from sympy import Matrix
from sympy.utilities.exceptions import SymPyDeprecationWarning
import warnings


warnings.filterwarnings("error", category=SymPyDeprecationWarning)
Matrix([[True]])


# I figure this out after reading multiple issues https://github.com/sympy/sympy/issues?q=is%3Aissue+is%3Aopen+SymPyDeprecationWarning of sympy.
# It would be good to know if this is documented somewhere or not.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

Commit:
private/python_ipc_native: Disable 'dbg_no_array' to keep in sync...

...with 'inst/private/python_header.py'.

Follow-up to 56bd71902ca0faf5fa88c1bc7fbeba8bf6f99ce7.

* inst/private/python_ipc_native.m: Disable it.

cbm755 committed:
Use S.Zero not integer 0

Fixes #1228.

cbm755 committed:
Array: fix a failing test in subasgn

cbm755 committed:
isconstant: fix behaviour with Array and add tests for errors

cbm755 committed:
levels=1 sufficient with tolist() from 2dArray/Matrix

cbm755 committed:
find: work with Array

cbm755 committed:
Fix some unused code

This was broken with syntax errors since 2016...  Apply fix for Array
but leave a comment that this is unused and untested.  It is referred to
in isprime but commented out.

Commit:
@sym/transpose: Use sympy function 'transpose'.

Follow-up to 136184d6df506bc9945d8f341e9b46691895c5b4.

* inst/@sym/transpose.m: Use it.

Commit:
@sym/vertcat: Use sympy function 'flatten'.

Follow-up to d3b154731803c95460af08722e7a62a718d20870.

* inst/@sym/vertcat.m: Use it.

Commit:
make_matrix_or_array: Generalise to accept an iterable of iterables.

* inst/private/{python_header.py,python_ipc_native.m}: Generalise
function 'make_matrix_or_array', use sympy function 'flatten' in
'make_matrix_or_array'.

Commit:
@sym/private/mat_rclist_asgn: Use sympy function 'flatten'.

* inst/@sym/private/mat_rclist_asgn.m: Use it.

Commit:
@sym/private/mat_replace: Make matrix mutable before modifying.

We did not run into issues in the past because many sympy functions
return a mutable matrix. However, 'transpose' used in
c9f6f0f9b1d0edaee5c29ce7d9e4ae2ece8708d4 returns an immutable
matrix. This change should avoid similar problems in the future.

Follow-up to c9f6f0f9b1d0edaee5c29ce7d9e4ae2ece8708d4.

* inst/@sym/private/mat_replace.m: Make it mutable.

cbm755 committed:
elementwise_op: call make_array_or_matrix

Instead of doing it ourselves, prepare a list of lists and call the
centralized helper function.

cbm755 committed:
Rough gating of Array on older SymPy

cbm755 committed:
make_2d_sym: support flat input and shape

Fixes #1236.  For now, just add a new underscore version that takes a
shape kwarg.

cbm755 committed:
Workaround nsolve with immutable x vars

cbm755 committed:
subs: after Array port was creating incorrect empties

Fixes #1236.  Or at least the part of #1236 that pertains to `subs`.
Added new tests.

cbm755 committed:
vertcat: tests for empty cases

cbm755 committed:
vertcat: use a flat list and shape for correct empty

Fixes #1236.

cbm755 committed:
horzcat/vertcat: add tests for horzcat as well

cbm755 committed:
Port helpers from list-of-lists to flat+shape

Related to #1236.  I don't see a problem with the list-of-lists
approach, but these are the only routine using it so we can delete some
code if we port them to flat + shape.

cbm755 committed:
isntance doesn't need special None handling
